util/passwordbasedauthentication.c

```diff
@@ -0,0 +1,271 @@
+      || is_prefix_not_supported (prefix))
+    {
+      output[0] = '*';
+      goto exit;
+    }
+  if (rbytes == NULL)
+    {
+      internal_rbytes = malloc (16);
+      if (get_random (internal_rbytes, 16) != 0)
+        {
+          output[0] = '*';
+          goto exit;
+        }
+      nrbytes = 16;
+      rbytes = internal_rbytes;
+    }
+  written = snprintf (output, output_size, "%srounds=%lu$",
+                      prefix == NULL ? PREFIX_DEFAULT : prefix, count);
+  while (written + 5 < (unsigned int) output_size
+         && used + 3 < (unsigned int) nrbytes && (used * 4 / 3) < 16)
+    {
+      value = ((unsigned long) rbytes[used + 0] << 0)
+              | ((unsigned long) rbytes[used + 1] << 8)
+              | ((unsigned long) rbytes[used + 2] << 16);
+      output[written] = ascii64[value & 0x3f];
+      output[written + 1] = ascii64[(value >> 6) & 0x3f];
+      output[written + 2] = ascii64[(value >> 12) & 0x3f];
+      output[written + 3] = ascii64[(value >> 18) & 0x3f];
+      written += 4;
+      used += 3;
+    }
+  output[written] = '\0';
+exit:
+  if (internal_rbytes != NULL)
+    free (internal_rbytes);
+  return output[0] == '*' ? 0 : output;
+}
+
+#endif
+
+struct PBASettings *
+pba_init (const char *pepper, unsigned int pepper_size, unsigned int count,
+          char *prefix)
+{
+  unsigned int i = 0;
+  struct PBASettings *result = NULL;
+  if (pepper_size > MAX_PEPPER_SIZE)
+    goto exit;
+  if (prefix != NULL && is_prefix_not_supported (prefix))
+    goto exit;
+  result = malloc (sizeof (struct PBASettings));
+  for (i = 0; i < MAX_PEPPER_SIZE; i++)
+    result->pepper[i] = pepper != NULL && i < pepper_size ? pepper[i] : 0;
+  result->count = count == 0 ? COUNT_DEFAULT : count;
+  result->prefix = prefix == NULL ? PREFIX_DEFAULT : prefix;
+exit:
+  return result;
+}
+
+void
+pba_finalize (struct PBASettings *settings)
+{
+  free (settings);
+}
+int
+pba_is_phc_compliant (const char *setting)
+{
+  if (setting == NULL)
+    {
+      return 0;
+    }
+  return strlen (setting) > 1 && setting[0] == '$';
+}
+
+char *
+pba_hash (struct PBASettings *setting, const char *password)
+{
+  char *result = NULL, *settings = NULL, *tmp, *rslt;
+  int i;
+  struct crypt_data *data = NULL;
+
+  if (!setting || !password)
+    goto exit;
+  if (is_prefix_not_supported (setting->prefix) != 0)
+    goto exit;
+  settings = malloc (CRYPT_GENSALT_OUTPUT_SIZE);
+  if (crypt_gensalt_r (setting->prefix, setting->count, NULL, 0, settings,
+                       CRYPT_GENSALT_OUTPUT_SIZE)
+      == NULL)
+    goto exit;
+  tmp = settings + strlen (settings) - 1;
+  for (i = MAX_PEPPER_SIZE - 1; i > -1; i--)
+    {
+      if (setting->pepper[i] != 0)
+        tmp[0] = setting->pepper[i];
+      tmp--;
+    }
+
+  data = calloc (1, sizeof (struct crypt_data));
+  rslt = crypt_r (password, settings, data);
+  if (rslt == NULL)
+    goto exit;
+  result = malloc (CRYPT_OUTPUT_SIZE);
+  memcpy (result, rslt, CRYPT_OUTPUT_SIZE);
```

Comment:
As discussed I am not sure if this might crash if crypt_r returns a smaller string here under some situations.

Reply:
changed to MIN(strlen(rslt), CRYPT_OUTPUT_SIZE)